Language/Haskell/GhcMod/ImportedFrom.hs

```diff
@@ -0,0 +1,826 @@
+    hasPostfixMatch xs s = last (separateBy '.' s) `elem` xs
+
+-- | The last thing with a single export must be the match? Iffy.
+getLastMatch :: [ModuleExports] -> Maybe ModuleExports
+getLastMatch exports = Safe.lastMay $ filter f exports
+  where
+    f me = length (qualifiedExports me) == 1
+
+-- | Try to look up the Haddock URL for a symbol.
+guessHaddockUrl
+    :: forall m.
+       (GhcMonad m, MonadIO m, GmOut m, GmLog m)
+    => ModSummary
+    -> FilePath
+    -> String
+    -> String
+    -> Int
+    -> Int
+    -> FilePath
+    -> (FilePath -> [String] -> String -> IO String)
+    -> [GhcPkgDb]
+    -> m String
+guessHaddockUrl modSum targetFile targetModule symbol lineNr colNr ghcPkg readProc pkgDbStack = do
+    gmLog GmDebug "guessHaddockUrl" $ strDoc $ "targetFile: "   ++ targetFile
+    gmLog GmDebug "guessHaddockUrl" $ strDoc $ "targetModule: " ++ targetModule
+    gmLog GmDebug "guessHaddockUrl" $ strDoc $ "symbol: "       ++ show symbol
+    gmLog GmDebug "guessHaddockUrl" $ strDoc $ "line nr: "      ++ show lineNr
+    gmLog GmDebug "guessHaddockUrl" $ strDoc $ "col nr: "       ++ show colNr
+
+    dflags <- GHC.getSessionDynFlags
+
+    let textualImports = ghc_ms_textual_imps modSum
+        importDecls0 = map (toImportDecl dflags) textualImports
+
+    gmLog GmDebug "guessHaddockUrl" $ strDoc $ "haskellModuleNames0: " ++ show importDecls0
+
+    -- If symbol is something like DM.lookup, then restrict importDecls0 to the
+    -- one that has modImportedAs == Just "DM".
+    let importDecls1 = filterMatchingQualifiedImport symbol importDecls0
+
+    -- If that filter left us with nothing, revert back to the original list.
+    let importDecls2 = if null importDecls1
+                                then importDecls0
+                                else importDecls1
+
+    qnames <- filter (not . (' ' `elem`)) <$> qualifiedName targetModule lineNr colNr symbol (map modName importDecls2) :: m [String]
+    gmLog GmDebug "guessHaddockUrl" $ strDoc $ "qnames: " ++ show qnames
+
+    let symbolToUse :: String
+        symbolToUse = case qnames of
+                        (qq:_) -> qq -- We got a qualified name, with qualified printing. Qualified!
+                        []     -> throw $ GMEString "ImportedFrom: qnames is empty."
+
+    gmLog GmDebug "guessHaddockUrl" $ strDoc $ "symbolToUse: " ++ symbolToUse
+
+    -- Sometimes we have to load an extra module (using setContext) otherwise
+    -- we can't look up the global reader environment without causing a GHC panic.
+    -- For example 'Int' comes from GHC.Types, which is picked up here via the
+    -- full qualified name.
+    let parsedPackagesAndQualNames = map (runRP parsePackageAndQualName) qnames
+
+        extraImportDecls :: [NiceImportDecl]
+        extraImportDecls = case Safe.headMay parsedPackagesAndQualNames of
+            Just (Right (_, x)) -> case moduleOfQualifiedName x of
+                                      Just x' -> [NiceImportDecl
```

Comment:
This huge whitespace block just looks ugly, can you pull the `NiceImportDecl` thing out into a binder in the enclosing let and use it in there instead please?